pip/pip-274.md

```diff
@@ -0,0 +1,126 @@
+
+
+For example, we need to parse the Kafka format then compact message in Kop, but the current implementation of Pulsar topic compaction does not support this feature.
+Another the topic compaction logic implemented in `TwoPhaseCompactor` only compacts messages to the last one, but sometimes we need to keep the first valid message e.g [`StrategicTwoPhaseCompactor`](https://github.com/coderzc/pulsar/blob/0e9935c493060b13b322a84c5418146423992369/pulsar-broker/src/main/java/org/apache/pulsar/compaction/StrategicTwoPhaseCompactor.java).
+
+So we need to make the topic compactor pluggable to support more compaction strategy.
+
+# Goals
+
+## In Scope
+
+<!--
+What this PIP intend to achieve once It's integrated into Pulsar.
+Why does it benefit Pulsar.
+-->
+
+Make the compactor pluggable.
+
+## Out of Scope
+
+<!--
+Describe what you have decided to keep out of scope, perhaps left for a different PIP/s.
+-->
+
+
+# High Level Design
+
+<!--
+Describe the design of your solution in *high level*.
+Describe the solution end to end, from a birds-eye view.
+Don't go into implementation details in this section.
+
+I should be able to finish reading from beginning of the PIP to here (including) and understand the feature and 
+how you intend to solve it, end to end.
+
+DON'T
+* Avoid code snippets, unless it's essential to explain your intent.
+-->
+
+Make the topic compactor pluggable, users can customize the compactor implementation according to their own special scenarios.
+
+
+# Detailed Design
+
+## Design & Implementation Details
+
+<!--
+This is the section where you dive into the details. It can be:
+* Concrete class names and their roles and responsibility, including methods.
+* Code snippets of existing code.
+* Interface names and its methods.
+* ...
+-->
+* Define a standard Compactor interface that specifies the methods and properties that the Compactor implementation needs to implement. This interface should include methods for Compactor initialization, Compactor execution, and getting Compactor stats.
+```java
+public interface Compactor {
+
+    void initialize(ServiceConfiguration conf,
+                    PulsarClient pulsar,
+                    BookKeeper bk,
+                    ScheduledExecutorService scheduler);
+
+    CompletableFuture<Long> compact(String topic);
+
+    CompactorMXBean getStats();
+}
+```
+
+* Rename `org.apache.pulsar.compaction.Compactor` to `org.apache.pulsar.compaction.AbstractCompactor` and make it implement `Compactor` interface.
+
+* Load custom compactor based on configuration in `org.apache.pulsar.broker.PulsarService.newCompactor` and `CompactorTool`.
+
+## Public-facing Changes
+
+<!--
+Describe the additions you plan to make for each public facing component. 
+Remove the sections you are not changing.
+Clearly mark any changes which are BREAKING backward compatability.
+-->
+
+
+### Configuration
```

Comment:
If I want `namespace A` to apply policy `TwoPhaseCompactor`, but `namespace B` applies policy `KafkaCompactor`, how can I configure it?

Reply:
Do not support the namespace level configuration. This is a global configuration.
By the way, since a topic may have Kafka Format and Pulsar Format at the same time, you need to be compatible with Pulsar Format in `KafkaCompactor`.

Reply:
If I have a Pulsar cluster and there are some topics on this cluster, and we need to migrate Kafka to Pulsar now, how can I do?

Reply:
>since a topic may have Kafka Format and Pulsar Format at the same time, you need to be compatible with Pulsar Format in KafkaCompactor.

Then, you copied the logic from the existing compactor? :( 
IMO, I wish we could support ns or topic-level configuration to avoid someday we will implement a powerful compactor. (one compactor supports all of protocol)

Reply:
> If I have a Pulsar cluster and there are some topics on this cluster, and we need to migrate Kafka to Pulsar now, how can I do?

For KafkaCompactor, you need to deal with Kafka Format and Pulsar Format, Kop can send Pulsar Format or Kafka Format

Reply:
> Then, you copied the logic from the existing compactor? :(

Yes, since a topic may be used by Kafka client and Pulsar client at the same time

Reply:
A consensus was reached after a discussion with @coderzc 
- This PIP only tries to make Compactor configurable, with no strong relation to KOP
- The new Compactor should be compatible with all the functions of `TwoPhaseCompactor`, it guarantees compatibility.

Reply:
Did you change the interface to support all functions of TwoPhaseCompactor?

Reply:
oh

Reply:
BTW, could you provide a link to explain the difference between Kafka Format and Pulsar Format? Whether a link to a document or a link to code is ok

Reply:
> Yes, since a topic may be used by Kafka client and Pulsar client at the same time

I'm afraid I still disagree with integrating all of the features in one plugin. This means all of the users want to implement a new compactor, and they also need to copy the existing logic into the new plugin. 

Could you consider adding more data protocols? you will introduce a new strategy in your implementation to make the logic more clear and readable. :(

Reply:
> The format rules of "Pulsar Format" and existing pulsar are exactly the same, right?

Yes

> KafkaCompactor can handle both Kafka Format and Pulsar Format, right?

Yes

> If not[2], when users want to migrate Kafka to an existing Pulsar cluster, how can they do it?

If `KafkacomPactor` is not implemented, you may need to use `Pulsar Format` in Kop to produce messages, and Kop will automatically transform `Kafka Format` to `Pulsar Format`.